### PR TITLE
[InstCombine] Fold zext(X) + C2 pred C -> X + C3 pred C4

### PR DESCRIPTION
**Pull request by dtcxzyw**
Motivating case from https://github.com/torvalds/linux/blob/9852d85ec9d492ebef56dc5f229416c925758edc/drivers/gpu/drm/drm_edid.c#L5238-L5240:
```
define i1 @src(i8 noundef %v13) {
entry:
  %conv1 = zext i8 %v13 to i32
  %add = add nsw i32 %conv1, -4
  %cmp = icmp ult i32 %add, 3
  %cmp4 = icmp slt i8 %v13, 4
  %cond = select i1 %cmp4, i1 true, i1 %cmp
  ret i1 %cond
}

define i1 @tgt(i8 noundef %v13) {
entry:
  %cmp4 = icmp slt i8 %v13, 7
  ret i1 %cmp4
}
```

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -3165,6 +3165,26 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
                         Builder.CreateAdd(X, ConstantInt::get(Ty, *C2 - C - 1)),
                         ConstantInt::get(Ty, ~C));
 
+  // zext(V) + C2 <u C -> V + trunc(C2) <u trunc(C) iff C2 s<0 && C s>0
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3168]: Think the 'iff ...' is pretty misleading given that you also need 'C2' and 'C' to fit in the new bitwidth.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -3168,6 +3168,31 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
...
                         ConstantInt::get(Ty, ~C));
 
+  // zext(V) + C2 pred C -> V + C3 pred' C4
+  Value *V;
+  if (match(X, m_ZExt(m_Value(V)))) {
+    Type *NewCmpTy = V->getType();
+    unsigned CmpBW = Ty->getScalarSizeInBits();
+    unsigned NewCmpBW = NewCmpTy->getScalarSizeInBits();
+    if (shouldChangeType(Ty, NewCmpTy)) {
+      if (auto ZExtCR = CR.exactIntersectWith(ConstantRange(
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3178]: We need a new ConstantRange API to convert ranges for `zext(V)` to ranges for `V`.
For example, we can convert `zext(i8 X to i32) - 255 u< -4` to `X + 1 u< -4`. But current implementation cannot achieve this.

**Comment by goldsteinn** [comment]
Is this intentionally still in draft form? Or are you ready for review?

**Comment by dtcxzyw** [comment]
> Is this intentionally still in draft form? Or are you ready for review?

It is ready for review now. I believe I cannot generalize it further without new ConstantRange API.

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -3169,6 +3169,31 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
...
+              APInt::getZero(CmpBW), APInt::getOneBitSet(CmpBW, NewCmpBW)))) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3180]: Can this condition be written as `CR.getActiveBits() <= NewCmpBW`? The exact intersection here seems pretty roundabout.

File: llvm/test/Transforms/InstCombine/icmp-add.ll [hunk 4]
```diff
@@ -3183,3 +3183,94 @@ define i1 @icmp_of_ucmp_plus_const_with_const(i32 %x, i32 %y) {
...
+; CHECK-NEXT:    [[ADD:%.*]] = add nsw i32 [[CONV]], -4
+; CHECK-NEXT:    [[CMP:%.*]] = icmp ult i32 [[ADD]], 3
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+entry:
+  %conv = zext i7 %x to i32
+  %add = add i32 %conv, -4
+  %cmp = icmp ult i32 %add, 3
+  ret i1 %cmp
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-add.ll:3276]: Negative test for the range check failing?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**